OpenRA.Mods.Common/Activities/HarvesterDockSequence.cs

```diff
@@ -62,10 +62,15 @@ public override bool Tick(Actor self)
 
 				case DockingState.Dock:
 					if (Refinery.IsInWorld && !Refinery.IsDead)
+					{
 						foreach (var nd in Refinery.TraitsImplementing<INotifyDocking>())
 							nd.Docked(Refinery, self);
 
-					OnStateDock(self);
+						OnStateDock(self);
+					}
+					else
+						dockingState = DockingState.Undock;
```

Comment:
Jumping to `Complete` instead of `Undock` improves the situation but may have other side effects.

Reply:
`Undock` is correct, as mentioned below I found a proper fix for this.